File: test/point_test.cpp
#include <limits>
#include <sstream>

#include <catch2/catch.hpp>
#include <copc-lib/geometry/box.hpp>
#include <copc-lib/geometry/vector3.hpp>
#include <copc-lib/las/point.hpp>
#include <copc-lib/las/points.hpp>

using namespace copc::las;
using namespace std;
using Catch::Matchers::WithinAbs;

Point PackAndUnpack(Point p, copc::Vector3 scale, copc::Vector3 offset)
{
    std::stringstream out;
    p.Pack(out, scale, offset);
    return *Point::Unpack(out, p.PointFormatId(), scale, offset, p.EbByteSize());
}

TEST_CASE("Point tests", "[Point]")
{
    SECTION("Point Format Initialization")
    {

        auto point = Point(6);

        REQUIRE(point.HasRgb() == false);
        REQUIRE(point.HasNir() == false);

        auto point_ext = Point(8);

        REQUIRE(point_ext.HasRgb() == true);
        REQUIRE(point_ext.HasNir() == true);
    }

    SECTION("Point Test")
    {
        auto point6 = Point(6);
        // Position
        point6.X(0.05);
        point6.Y(1.11);
        point6.Z(-9.99);
        REQUIRE_THAT(point6.X(), WithinAbs(0.05, 0.000001));
        REQUIRE_THAT(point6.Y(), WithinAbs(1.11, 0.000001));
        REQUIRE_THAT(point6.Z(), WithinAbs(-9.99, 0.000001));

        // Intensity
        point6.Intensity(std::numeric_limits<uint16_t>::max());
        REQUIRE(point6.Intensity() == std::numeric_limits<uint16_t>::max());

        // Return BitField
        point6.ReturnsBitField(std::numeric_limits<uint8_t>::max());
        REQUIRE(point6.ReturnsBitField() == std::numeric_limits<uint8_t>::max());

        // Return Number
        point6.ReturnNumber(0);
        REQUIRE(point6.ReturnNumber() == 0);
        point6.ReturnNumber(15);
        REQUIRE(point6.ReturnNumber() == 15);
        REQUIRE_THROWS(point6.ReturnNumber(16));

        // Number return
        point6.NumberOfReturns(0);
        REQUIRE(point6.NumberOfReturns() == 0);
        point6.NumberOfReturns(15);
        REQUIRE(point6.NumberOfReturns() == 15);
        REQUIRE_THROWS(point6.NumberOfReturns(16));

        // Flags
        point6.FlagsBitField(std::numeric_limits<uint8_t>::max());
        REQUIRE(point6.FlagsBitField() == std::numeric_limits<uint8_t>::max());

        // Synthetic
        point6.Synthetic(false);
        REQUIRE(point6.Synthetic() == false);
        point6.Synthetic(true);
        REQUIRE(point6.Synthetic() == true);

        // KeyPoint
        point6.Withheld(false);
        REQUIRE(point6.Withheld() == false);
        point6.Withheld(true);
        REQUIRE(point6.Withheld() == true);

        // Withheld
        point6.Withheld(false);
        REQUIRE(point6.Withheld() == false);
        point6.Withheld(true);
        REQUIRE(point6.Withheld() == true);

        // Overlap
        point6.Overlap(false);
        REQUIRE(point6.Overlap() == false);
        point6.Overlap(true);
        REQUIRE(point6.Overlap() == true);

        // Scanner Channel
        point6.ScannerChannel(0);
        REQUIRE(point6.ScannerChannel() == 0);
        point6.ScannerChannel(3);
        REQUIRE(point6.ScannerChannel() == 3);
        REQUIRE_THROWS(point6.ScannerChannel(4));

        // Scan Direction
        point6.ScanDirectionFlag(true);
        REQUIRE(point6.ScanDirectionFlag() == true);
        point6.ScanDirectionFlag(false);
        REQUIRE(point6.ScanDirectionFlag() == false);

        // Edge of Flight Line
        point6.EdgeOfFlightLineFlag(true);
        REQUIRE(point6.EdgeOfFlightLineFlag() == true);
        point6.EdgeOfFlightLineFlag(false);
        REQUIRE(point6.EdgeOfFlightLineFlag() == false);

        // Classification
        point6.Classification(std::numeric_limits<uint8_t>::max());
        REQUIRE(point6.Classification() == std::numeric_limits<uint8_t>::max());
        point6.Classification(0);
        REQUIRE(point6.Classification() == 0);

        // Scan Angle
        point6.ScanAngle(-30000);
        REQUIRE(point6.ScanAngle() == -30000);
        REQUIRE(point6.ScanAngleDegrees() == -180.0);
        point6.ScanAngle(30000);
        REQUIRE(point6.ScanAngle() == 30000);
        REQUIRE(point6.ScanAngleDegrees() == 180.0);

        // User Data
        point6.UserData(std::numeric_limits<uint8_t>::max());
        REQUIRE(point6.UserData() == std::numeric_limits<uint8_t>::max());

        // Point Source ID
        point6.PointSourceId(std::numeric_limits<uint16_t>::max());
        REQUIRE(point6.PointSourceId() == std::numeric_limits<uint16_t>::max());

        // GPS Time
        point6.GPSTime(std::numeric_limits<double>::max());
        REQUIRE(point6.GPSTime() == std::numeric_limits<double>::max());

        // Point Record Length
        REQUIRE(point6.PointRecordLength() == 30);

        // Checks
        REQUIRE_THROWS(point6.Rgb(std::numeric_limits<uint16_t>::max(), std::numeric_limits<uint16_t>::max(),
                                  std::numeric_limits<uint16_t>::max()));
        REQUIRE_THROWS(point6.Red());
        REQUIRE_THROWS(point6.Green());
        REQUIRE_THROWS(point6.Blue());
        REQUIRE_THROWS(point6.Nir(std::numeric_limits<uint16_t>::max()));
        REQUIRE_THROWS(point6.Nir());
        REQUIRE_NOTHROW(point6.ToString());

        auto point7 = Point(7);

        point7.Rgb(std::numeric_limits<uint16_t>::max(), std::numeric_limits<uint16_t>::max(),
                   std::numeric_limits<uint16_t>::max());
        REQUIRE(point7.Red() == std::numeric_limits<uint16_t>::max());
        REQUIRE(point7.Green() == std::numeric_limits<uint16_t>::max());
        REQUIRE(point7.Blue() == std::numeric_limits<uint16_t>::max());
        REQUIRE(point7.PointRecordLength() == 36);

        REQUIRE_THROWS(point7.Nir(std::numeric_limits<uint16_t>::max()));
        REQUIRE_THROWS(point7.Nir());

        auto point8 = Point(8);

        point8.Nir(std::numeric_limits<uint16_t>::max());
        REQUIRE(point8.Nir() == std::numeric_limits<uint16_t>::max());
        REQUIRE(point8.PointRecordLength() == 38);
    }

    SECTION("Point format conversion")
    {
        auto point = Point(6);

        REQUIRE(point.GPSTime() == 0);
        REQUIRE(point.ScannerChannel() == 0);
        REQUIRE(point.Overlap() == 0);
        REQUIRE(point.GPSTime() == 0);

        point.ToPointFormat(7);
        REQUIRE(point.GPSTime() == 0);
        REQUIRE(point.ScannerChannel() == 0);
        REQUIRE(point.Overlap() == 0);
        REQUIRE(point.GPSTime() == 0);
        REQUIRE(point.Red() == 0);
        REQUIRE(point.Green() == 0);
        REQUIRE(point.Blue() == 0);

        point.ToPointFormat(8);
        REQUIRE(point.GPSTime() == 0);
        REQUIRE(point.ScannerChannel() == 0);
        REQUIRE(point.Overlap() == 0);
        REQUIRE(point.GPSTime() == 0);
        REQUIRE(point.Red() == 0);
        REQUIRE(point.Green() == 0);
        REQUIRE(point.Blue() == 0);
        REQUIRE(point.Nir() == 0);
    }

    SECTION("Point conversion")
    {
        auto point = Point(6);

        point.ReturnNumber(5);
        point.NumberOfReturns(6);
        point.ScanDirectionFlag(false);
        point.EdgeOfFlightLineFlag(true);
        point.Classification(15);
        point.Synthetic(true);
        point.KeyPoint(false);
        point.Withheld(true);
        point.ScanAngleDegrees(45);

        point.ToPointFormat(7);

        REQUIRE(point.ReturnNumber() == 5);
        REQUIRE(point.NumberOfReturns() == 6);
        REQUIRE(point.ScanDirectionFlag() == false);
        REQUIRE(point.EdgeOfFlightLineFlag() == true);
        REQUIRE(point.Classification() == 15);
        REQUIRE(point.Synthetic() == true);
        REQUIRE(point.KeyPoint() == false);
        REQUIRE(point.Withheld() == true);
        REQUIRE(point.Overlap() == false);
        REQUIRE(point.ScanAngleDegrees() == 45);
        REQUIRE(point.ScannerChannel() == 0);
        REQUIRE(point.Red() == 0);
        REQUIRE(point.Green() == 0);
        REQUIRE(point.Blue() == 0);
        REQUIRE_THROWS(point.Nir());

        point.Red(150);
        point.Green(200);
        point.Blue(250);

        point.ToPointFormat(8);

        REQUIRE(point.ReturnNumber() == 5);
        REQUIRE(point.NumberOfReturns() == 6);
        REQUIRE(point.ScanDirectionFlag() == false);
        REQUIRE(point.EdgeOfFlightLineFlag() == true);
        REQUIRE(point.Classification() == 15);
        REQUIRE(point.Synthetic() == true);
        REQUIRE(point.KeyPoint() == false);
        REQUIRE(point.Withheld() == true);
        REQUIRE(point.Overlap() == false);
        REQUIRE(point.ScanAngleDegrees() == 45);
        REQUIRE(point.ScannerChannel() == 0);
        REQUIRE(point.Red() == 150);
        REQUIRE(point.Green() == 200);
        REQUIRE(point.Blue() == 250);
        REQUIRE(point.Nir() == 0);
    }

    SECTION("Operator == and !=")
    {

        // Format 6
        auto point = Point(6);
        auto point_other = Point(6);
        REQUIRE(point == point_other);

        // Format 7
        point_other.ToPointFormat(7);
        REQUIRE(point != point_other);
        point.ToPointFormat(7);
        REQUIRE(point == point_other);

        // Format 8
        point_other.ToPointFormat(8);
        REQUIRE(point != point_other);
        point.ToPointFormat(8);
        REQUIRE(point == point_other);

        // Atributes

        point.X(4);
        REQUIRE(point != point_other);
        point_other.X(4);
        REQUIRE(point == point_other);

        point.Y(4);
        REQUIRE(point != point_other);
        point_other.Y(4);
        REQUIRE(point == point_other);

        point.Z(4);
        REQUIRE(point != point_other);
        point_other.Z(4);
        REQUIRE(point == point_other);

        point.Intensity(4);
        REQUIRE(point != point_other);
        point_other.Intensity(4);
        REQUIRE(point == point_other);

        point.ReturnNumber(4);
        REQUIRE(point != point_other);
        point_other.ReturnNumber(4);
        REQUIRE(point == point_other);

        point.NumberOfReturns(4);
        REQUIRE(point != point_other);
        point_other.NumberOfReturns(4);
        REQUIRE(point == point_other);

        point.Classification(4);
        REQUIRE(point != point_other);
        point_other.Classification(4);
        REQUIRE(point == point_other);

        point.ScanDirectionFlag(true);
        REQUIRE(point != point_other);
        point_other.ScanDirectionFlag(true);
        REQUIRE(point == point_other);

        point.EdgeOfFlightLineFlag(true);
        REQUIRE(point != point_other);
        point_other.EdgeOfFlightLineFlag(true);
        REQUIRE(point == point_other);

        point.Synthetic(true);
        REQUIRE(point != point_other);
        point_other.Synthetic(true);
        REQUIRE(point == point_other);

        point.KeyPoint(true);
        REQUIRE(point != point_other);
        point_other.KeyPoint(true);
        REQUIRE(point == point_other);

        point.Withheld(true);
        REQUIRE(point != point_other);
        point_other.Withheld(true);
        REQUIRE(point == point_other);

        point.Overlap(true);
        REQUIRE(point != point_other);
        point_other.Overlap(true);
        REQUIRE(point == point_other);

        point.ScannerChannel(2);
        REQUIRE(point != point_other);
        point_other.ScannerChannel(2);
        REQUIRE(point == point_other);

        point.UserData(4);
        REQUIRE(point != point_other);
        point_other.UserData(4);
        REQUIRE(point == point_other);

        point.PointSourceId(4);
        REQUIRE(point != point_other);
        point_other.PointSourceId(4);
        REQUIRE(point == point_other);

        point.GPSTime(4.0);
        REQUIRE(point != point_other);
        point_other.GPSTime(4.0);
        REQUIRE(point == point_other);

        point.Red(4);
        REQUIRE(point != point_other);
        point_other.Red(4);
        REQUIRE(point == point_other);

        point.Green(4);
        REQUIRE(point != point_other);
        point_other.Green(4);
        REQUIRE(point == point_other);

        point.Blue(4);
        REQUIRE(point != point_other);
        point_other.Blue(4);
        REQUIRE(point == point_other);

        point.Nir(4);
        REQUIRE(point != point_other);
        point_other.Nir(4);
        REQUIRE(point == point_other);
    }

    SECTION("Point ExtraByte")
    {
        auto point = Point(6);
        REQUIRE(point.PointFormatId() == 6);
        REQUIRE(point.PointRecordLength() == 30);
        REQUIRE_THROWS(point.ExtraBytes(std::vector<uint8_t>{2, 3, 4}));
        REQUIRE(point.ExtraBytes().empty());
        REQUIRE(point.EbByteSize() == 0);

        point = Point(6, 5);
        REQUIRE(point.PointFormatId() == 6);
        REQUIRE(point.PointRecordLength() == 30 + 5);
        REQUIRE(point.EbByteSize() == 5);
        REQUIRE(point.ExtraBytes().size() == 5);
        REQUIRE_THROWS(point.ExtraBytes(std::vector<uint8_t>{2, 3, 4}));
        REQUIRE_NOTHROW(point.ExtraBytes(std::vector<uint8_t>{2, 3, 4, 5, 6}));
        REQUIRE(point.ExtraBytes() == std::vector<uint8_t>{2, 3, 4, 5, 6});
    }

    SECTION("Operator =")
    {
        auto point = Point(8, 2);

        point.X(4);
        point.Y(4);
        point.Z(4);

        point.GPSTime(4.0);
        point.Red(4.0);
        point.Nir(4.0);

        point.ExtraBytes(std::vector<uint8_t>{2, 5});

        auto point_other = Point(6);
        point_other = point;

        REQUIRE(point == point_other);
    }

    SECTION("Packing and Unpacking")
    {

        std::stringstream ss;
        auto scale = copc::Vector3::DefaultScale();
        auto offset = copc::Vector3::DefaultOffset();
        auto orig_point = Point(6, 2); // use two extra bytes
        orig_point.X(20);
        orig_point.Y(-20);
        orig_point.Z(100000);
        orig_point.ScanAngle(2000);

        // Format 6
        auto point = orig_point;
        point.Pack(ss, scale, offset);
        auto point_other =
            *Point::Unpack(ss, 6, copc::Vector3::DefaultScale(), copc::Vector3::DefaultOffset(), point.EbByteSize());
        REQUIRE(point == point_other);

        // Format 7
        point.ToPointFormat(7);
        point.Pack(ss, scale, offset);
        point_other =
            *Point::Unpack(ss, 7, copc::Vector3::DefaultScale(), copc::Vector3::DefaultOffset(), point.EbByteSize());
        REQUIRE(point == point_other);

        // Format 8
        point.ToPointFormat(8);
        point.Pack(ss, scale, offset);
        point_other =
            *Point::Unpack(ss, 8, copc::Vector3::DefaultScale(), copc::Vector3::DefaultOffset(), point.EbByteSize());
        REQUIRE(point == point_other);
    }

    SECTION("Scaled XYZ")
    {
        int8_t pfid = 8;

        SECTION("No scale and offset")
        {
            auto point = Point(pfid);

            point.X(4);
            point.Y(4.5);
            point.Z(-0.1);

            auto test_point = PackAndUnpack(point, copc::Vector3(1, 1, 1), copc::Vector3(0, 0, 0));

            REQUIRE(test_point.X() == 4);
            REQUIRE(test_point.Y() == 5);
            REQUIRE(test_point.Z() == 0);
        }
        SECTION("Scale test")
        {
            auto point = Point(pfid);

            point.X(0.01);
            point.Y(0.02);
            point.Z(0.03);

            auto test_point = PackAndUnpack(point, copc::Vector3(1, 1, 1), copc::Vector3(0, 0, 0));

            REQUIRE(test_point.X() == 0);
            REQUIRE(test_point.Y() == 0);
            REQUIRE(test_point.Z() == 0);

            test_point = PackAndUnpack(point, copc::Vector3(0.01, 0.01, 0.01), copc::Vector3(0, 0, 0));

            REQUIRE(test_point.X() == 0.01);
            REQUIRE(test_point.Y() == 0.02);
            REQUIRE(test_point.Z() == 0.03);
        }
        SECTION("Offset test")
        {
            copc::Vector3 scale = {1, 1, 1};
            copc::Vector3 offset = {50001.456, 4443.123, -255.001};
            auto point = Point(pfid);

            point.X(50502.888);
            point.Y(4002.111);
            point.Z(-80.5);

            auto test_point = PackAndUnpack(point, scale, offset);
            REQUIRE_THAT(test_point.X(), WithinAbs(50502.456, 0.000001));
            REQUIRE_THAT(test_point.Y(), WithinAbs(4002.123, 0.000001));
            REQUIRE_THAT(test_point.Z(), WithinAbs(-80.001, 0.000001));
        }
        SECTION("Scale and Offset test")
        {
            copc::Vector3 scale = {0.001, 0.001, 0.001};
            copc::Vector3 offset = {50001.456, 4443.123, -255.001};
            auto point = Point(pfid);

            point.X(50502.888);
            point.Y(4002.111);
            point.Z(-80.5);

            auto test_point = PackAndUnpack(point, scale, offset);
            REQUIRE_THAT(test_point.X(), WithinAbs(50502.888, 0.000001));
            REQUIRE_THAT(test_point.Y(), WithinAbs(4002.111, 0.000001));
            REQUIRE_THAT(test_point.Z(), WithinAbs(-80.5, 0.000001));
        }
        SECTION("Change Scale test")
        {
            copc::Vector3 offset = {50001.456, 4443.123, -255.001};
            copc::Vector3 scale = {0.001, 0.001, 0.001};
            copc::Vector3 new_scale = {1, 1, 1};
            // This tests the case that we want to change the point's scale when we pack it
            auto point = Point(pfid);

            point.X(50502.888);
            point.Y(4002.111);
            point.Z(-80.5);

            // (value * scale) + offset
            REQUIRE_THAT(point.X(), WithinAbs(50502.888, 0.000001));
            REQUIRE_THAT(point.Y(), WithinAbs(4002.111, 0.000001));
            REQUIRE_THAT(point.Z(), WithinAbs(-80.5, 0.000001));

            // When we pack the points, that's when we unscale the internally stored double.
            // so we end up with:
            // UnscaledX = static_cast<T>(std::round((50502.888 - 50001.456) / 1)) = 501
            // UnscaledY = static_cast<T>(std::round((4002.111 - 4443.123) / 1)) = -441
            // UnscaledZ = static_cast<T>(std::round((-80.5 - -255.001) / 1)) = 175

            // Then when we unpack the points, we do the inverse:
            // (value * scale) + offset
            // X = 501 * 1 + 50001.456 = 50502.456
            // Y = -441 * 1 + 4443.123 = 4002.123
            // Z = 175 * 1 + -255.001 = -80.001
            auto test_point = PackAndUnpack(point, new_scale, offset);

            REQUIRE_THAT(test_point.X(), WithinAbs(50502.456, 0.0001));
            REQUIRE_THAT(test_point.Y(), WithinAbs(4002.123, 0.0001));
            REQUIRE_THAT(test_point.Z(), WithinAbs(-80.001, 0.0001));
        }
        // todo: add change offset and change offset/scale tests
        SECTION("Precision checks")
        {
            {
                auto point = Point(pfid);
                point.X(50501132.888123);
                std::stringstream out;
                REQUIRE_THROWS(point.Pack(out, {0.000001, 0.000001, 0.000001}, {0, 0, 0}));
            }
            {
                auto point = Point(pfid);
                point.X(0);
                std::stringstream out;
                REQUIRE_THROWS(point.Pack(out, {1, 1, 1}, {-8001100065, 0, 0}));
            }
        }
    }
    SECTION("Within")
    {
        auto point = Point(6);

        point.X(5);
        point.Y(5);
        point.Z(5);

        REQUIRE(point.Within(copc::Box::MaxBox()));
        REQUIRE(!point.Within(copc::Box::EmptyBox()));

        // 2D box
        REQUIRE(point.Within(copc::Box(0, 0, 5, 5)));
        REQUIRE(!point.Within(copc::Box(6, 0, 10, 5)));

        // 3D box
        REQUIRE(point.Within(copc::Box(0, 0, 0, 5, 5, 5)));
        REQUIRE(!point.Within(copc::Box(6, 0, 0, 10, 5, 5)));
    }
}

File: test/point_test.py
import sys

import copclib as copc
import pytest


def pack_and_unpack(p, scale, offset):
    points = copc.Points(p.point_format_id, p.EbByteSize())
    points.AddPoint(p)
    return copc.Points.Unpack(points.Pack(scale, offset), p.point_format_id, p.EbByteSize(), scale, offset)[0]

def test_constructor():
    point = copc.Points(
        point_format_id=6,
        eb_byte_size=0,
    ).CreatePoint()

    assert point.HasRgb() is False
    assert point.HasNir() is False

    point_ext = copc.Points(
        8
    ).CreatePoint()

    assert point_ext.HasRgb() is True
    assert point_ext.HasNir() is True


def test_point():
    point6 = copc.Points(
        6
    ).CreatePoint()
    # Position
    point6.x = 2147483647
    point6.y = 2147483647
    point6.z = 2147483647
    assert point6.x == 2147483647
    assert point6.y == 2147483647
    assert point6.z == 2147483647

    # intensity
    point6.intensity = 65535
    assert point6.intensity == 65535

    # Return BitField
    point6.returns_bit_field = 255
    assert point6.returns_bit_field == 255

    # Return Number
    point6.return_number = 0
    assert point6.return_number == 0
    point6.return_number = 15
    assert point6.return_number == 15
    with pytest.raises(RuntimeError):
        point6.return_number = 16

    # Number return
    point6.number_of_returns = 0
    assert point6.number_of_returns == 0
    point6.number_of_returns = 15
    assert point6.number_of_returns == 15
    with pytest.raises(RuntimeError):
        point6.number_of_returns = 16

    # Flags
    point6.flags_bit_field = 255
    assert point6.flags_bit_field == 255

    # synthetic
    point6.synthetic = False
    assert point6.synthetic is False
    point6.synthetic = True
    assert point6.synthetic is True

    # key_point
    point6.key_point = False
    assert point6.key_point is False
    point6.key_point = True
    assert point6.key_point is True

    # withheld
    point6.withheld = False
    assert point6.withheld is False
    point6.withheld = True
    assert point6.withheld is True

    # overlap
    point6.overlap = False
    assert point6.overlap is False
    point6.overlap = True
    assert point6.overlap is True

    # Scanner Channel
    point6.scanner_channel = 0
    assert point6.scanner_channel == 0
    point6.scanner_channel = 3
    assert point6.scanner_channel == 3
    with pytest.raises(RuntimeError):
        point6.scanner_channel = 4

    # Scan Direction
    point6.scan_direction_flag = True
    assert point6.scan_direction_flag is True
    point6.scan_direction_flag = False
    assert point6.scan_direction_flag is False

    # Edge of Flight Line
    point6.edge_of_flight_line = True
    assert point6.edge_of_flight_line is True
    point6.edge_of_flight_line = False
    assert point6.edge_of_flight_line is False

    # classification
    point6.classification = 255
    assert point6.classification == 255
    point6.classification = 0
    assert point6.classification == 0

    # Scan Angle
    point6.scan_angle = -30000
    assert point6.scan_angle == -30000
    assert point6.scan_angle_degrees == pytest.approx(-180.0)
    point6.scan_angle = 30000
    assert point6.scan_angle == 30000
    assert point6.scan_angle_degrees == pytest.approx(180.0)

    # User Data
    point6.user_data = 255
    assert point6.user_data == 255

    # Point Source ID
    point6.point_source_id = 65535
    assert point6.point_source_id == 65535

    # GPS Time
    point6.gps_time = sys.float_info.max
    assert point6.gps_time == sys.float_info.max

    # Point Record Length
    assert point6.point_record_length == 30

    # Checks
    with pytest.raises(RuntimeError):
        point6.rgb = [65535, 65535, 65535]
        assert point6.red
        assert point6.green
        assert point6.blue
        point6.nir = 65535
        assert point6.nir
        assert point6.scan_angleRank
        point6.scan_angleRank = 127
        point6.ReturnsScanDirEofBitFields = 255
        assert point6.ReturnsScanDirEofBitFields
        point6.classificationBitFields = 255
        assert point6.classificationBitFields

    # ToString
    str(point6)

    point7 = copc.Points(
        7
    ).CreatePoint()

    point7.rgb = [65535, 65535, 65535]
    assert point7.red == 65535
    assert point7.green == 65535
    assert point7.blue == 65535
    assert point7.point_record_length == 36

    with pytest.raises(RuntimeError):
        point7.nir = 65535
        assert point7.nir

    point8 = copc.Points(
        8
    ).CreatePoint()

    point8.nir = 65535
    assert point8.nir == 65535
    assert point8.point_record_length == 38


def test_point_conversion():
    point = copc.Points(
        6
    ).CreatePoint()

    assert point.gps_time == 0
    assert point.scanner_channel == 0
    assert point.overlap == 0
    assert point.gps_time == 0

    point.ToPointFormat(7)
    assert point.gps_time == 0
    assert point.scanner_channel == 0
    assert point.overlap == 0
    assert point.gps_time == 0
    assert point.red == 0
    assert point.green == 0
    assert point.blue == 0

    point.ToPointFormat(8)
    assert point.gps_time == 0
    assert point.scanner_channel == 0
    assert point.overlap == 0
    assert point.gps_time == 0
    assert point.red == 0
    assert point.green == 0
    assert point.blue == 0
    assert point.nir == 0


def test_operators_equal():
    # Format 0
    point = copc.Points(
        6
    ).CreatePoint()
    point_other = copc.Points(
        6
    ).CreatePoint()
    assert point == point_other

    # Format 7
    point_other.ToPointFormat(7)
    assert point != point_other
    point.ToPointFormat(7)
    assert point == point_other

    # Format 8
    point_other.ToPointFormat(8)
    assert point != point_other
    point.ToPointFormat(8)
    assert point == point_other

    # Atributes

    point.x = 4
    assert point != point_other
    point_other.x = 4
    assert point == point_other

    point.y = 4
    assert point != point_other
    point_other.y = 4
    assert point == point_other

    point.z = 4
    assert point != point_other
    point_other.z = 4
    assert point == point_other

    point.intensity = 4
    assert point != point_other
    point_other.intensity = 4
    assert point == point_other

    point.return_number = 4
    assert point != point_other
    point_other.return_number = 4
    assert point == point_other

    point.number_of_returns = 4
    assert point != point_other
    point_other.number_of_returns = 4
    assert point == point_other

    point.classification = 4
    assert point != point_other
    point_other.classification = 4
    assert point == point_other

    point.scan_direction_flag = True
    assert point != point_other
    point_other.scan_direction_flag = True
    assert point == point_other

    point.edge_of_flight_line = True
    assert point != point_other
    point_other.edge_of_flight_line = True
    assert point == point_other

    point.synthetic = True
    assert point != point_other
    point_other.synthetic = True
    assert point == point_other

    point.key_point = True
    assert point != point_other
    point_other.key_point = True
    assert point == point_other

    point.withheld = True
    assert point != point_other
    point_other.withheld = True
    assert point == point_other

    point.overlap = True
    assert point != point_other
    point_other.overlap = True
    assert point == point_other

    point.scanner_channel = 2
    assert point != point_other
    point_other.scanner_channel = 2
    assert point == point_other

    point.user_data = 4
    assert point != point_other
    point_other.user_data = 4
    assert point == point_other

    point.point_source_id = 4
    assert point != point_other
    point_other.point_source_id = 4
    assert point == point_other

    point.gps_time = 4.0
    assert point != point_other
    point_other.gps_time = 4.0
    assert point == point_other

    point.red = 4
    assert point != point_other
    point_other.red = 4
    assert point == point_other

    point.green = 4
    assert point != point_other
    point_other.green = 4
    assert point == point_other

    point.blue = 4
    assert point != point_other
    point_other.blue = 4
    assert point == point_other

    point.nir = 4
    assert point != point_other
    point_other.nir = 4
    assert point == point_other


def test_extra_byte():
    point = copc.Points(
        6
    ).CreatePoint()
    assert point.point_format_id == 6
    assert point.point_record_length == 30
    with pytest.raises(RuntimeError):
        point.extra_bytes = [2, 3, 4]
    assert len(point.extra_bytes) == 0
    assert point.EbByteSize() == 0

    point = copc.Points(
        6,
        eb_byte_size=5,
    ).CreatePoint()
    assert point.point_format_id == 6
    assert point.point_record_length == 30 + 5
    assert point.EbByteSize() == 5
    assert len(point.extra_bytes) == 5
    with pytest.raises(RuntimeError):
        point.extra_bytes = [2, 3, 4]
    point.extra_bytes = [2, 3, 4, 5, 6]
    assert point.extra_bytes == [2, 3, 4, 5, 6]


def test_operator_copy():
    point = copc.Points(
        8,
        eb_byte_size=2,
    ).CreatePoint()

    point.x = 4
    point.y = 4
    point.z = 4

    point.gps_time = 4.0
    point.red = 4
    point.nir = 4

    point.extra_bytes = [2, 5]

    point_other = point

    assert point == point_other


def test_scaled_xyz():

    pfid = 8

    # No scale and offset
    point = copc.Points(pfid).CreatePoint()

    point.x = 4
    point.y = 4.5
    point.z = -0.1

    test_point = pack_and_unpack(point, [1,1,1], [0,0,0])
    assert test_point.x == 4
    assert test_point.y == 5
    assert test_point.z == 0

    # Scale test
    point = copc.Points(
        pfid
    ).CreatePoint()

    point.x = 0.01
    point.y = 0.02
    point.z = 0.03

    test_point = pack_and_unpack(point, [1,1,1], [0,0,0])
    assert test_point.x == 0
    assert test_point.y == 0
    assert test_point.z == 0

    test_point = pack_and_unpack(point, [0.01,0.01,0.01], [0,0,0])

    assert test_point.x == 0.01
    assert test_point.y == 0.02
    assert test_point.z == 0.03

    # Offset test

    scale = [1, 1, 1]
    offset = copc.Vector3([50001.456, 4443.123, -255.001])
    point = copc.Points(pfid).CreatePoint()

    point.x = 50502.888
    point.y = 4002.111
    point.z = -80.5

    test_point = pack_and_unpack(point, scale, offset)

    assert test_point.x == pytest.approx(50502.456, 0.000001)
    assert test_point.y == pytest.approx(4002.123, 0.000001)
    assert test_point.z == pytest.approx(-80.001, 0.000001)

    # Scale and Offset test
    scale=[0.001, 0.001, 0.001]
    offset=[50001.456, 4443.123, -255.001]
    point = copc.Points(pfid).CreatePoint()

    point.x = 50502.888
    point.y = 4002.111
    point.z = -80.5

    test_point = pack_and_unpack(point, scale, offset)
    assert test_point.x == pytest.approx(50502.888, 0.000001)
    assert test_point.y == pytest.approx(4002.111, 0.000001)
    assert test_point.z == pytest.approx(-80.5, 0.000001)

    # Change scale test
    offset = [50001.456, 4443.123, -255.001]
    scale = [0.001, 0.001, 0.001]
    new_scale = [1,1,1]
    point = copc.Points(pfid).CreatePoint()

    point.x = 50502.888
    point.y = 4002.111
    point.z = -80.5

    test_point = pack_and_unpack(point, new_scale, offset)
    assert test_point.x == pytest.approx(50502.456, 0.000001)
    assert test_point.y == pytest.approx(4002.123, 0.000001)
    assert test_point.z == pytest.approx(-80.001, 0.000001)


    # Precision checks
    points = copc.Points(pfid)
    point = points.CreatePoint()
    point.x = 50501132.888123
    points.AddPoint(point)
    with pytest.raises(RuntimeError):
        points.Pack([0.000001, 0.000001, 0.000001], [0, 0, 0])


    points = copc.Points(pfid)
    point = points.CreatePoint()
    point.x = 0
    points.AddPoint(point)
    with pytest.raises(RuntimeError):
        points.Pack([1, 1, 1], [-8001100065, 0, 0])


def test_within():
    point = copc.Points(6).CreatePoint()

    point.x = 5
    point.y = 5
    point.z = 5

    assert point.Within(copc.Box.MaxBox())
    assert not point.Within(copc.Box.EmptyBox())

    # 2D box
    assert point.Within(copc.Box(0, 0, 5, 5))
    assert not point.Within(copc.Box(6, 0, 10, 5))

    # 3D box
    assert point.Within(copc.Box(0, 0, 0, 5, 5, 5))
    assert not point.Within(copc.Box(6, 0, 0, 10, 5, 5))
